<!DOCTYPE html>
<html>

<head>

<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Doublesex Project- Step by step methods using _Acromyrmex echinatior_ data</title>


<style type="text/css">
body {
  font-family: Helvetica, arial, sans-serif;
  font-size: 14px;
  line-height: 1.6;
  padding-top: 10px;
  padding-bottom: 10px;
  background-color: white;
  padding: 30px; }

body > *:first-child {
  margin-top: 0 !important; }
body > *:last-child {
  margin-bottom: 0 !important; }

a {
  color: #4183C4; }
a.absent {
  color: #cc0000; }
a.anchor {
  display: block;
  padding-left: 30px;
  margin-left: -30px;
  cursor: pointer;
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0; }

h1, h2, h3, h4, h5, h6 {
  margin: 20px 0 10px;
  padding: 0;
  font-weight: bold;
  -webkit-font-smoothing: antialiased;
  cursor: text;
  position: relative; }

h1:hover a.anchor, h2:hover a.anchor, h3:hover a.anchor, h4:hover a.anchor, h5:hover a.anchor, h6:hover a.anchor {
  background: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA09pVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMy1jMDExIDY2LjE0NTY2MSwgMjAxMi8wMi8wNi0xNDo1NjoyNyAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIiB4bWxuczp4bXBNTT0iaHR0cDovL25zLmFkb2JlLmNvbS94YXAvMS4wL21tLyIgeG1sbnM6c3RSZWY9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9zVHlwZS9SZXNvdXJjZVJlZiMiIHhtcDpDcmVhdG9yVG9vbD0iQWRvYmUgUGhvdG9zaG9wIENTNiAoMTMuMCAyMDEyMDMwNS5tLjQxNSAyMDEyLzAzLzA1OjIxOjAwOjAwKSAgKE1hY2ludG9zaCkiIHhtcE1NOkluc3RhbmNlSUQ9InhtcC5paWQ6OUM2NjlDQjI4ODBGMTFFMTg1ODlEODNERDJBRjUwQTQiIHhtcE1NOkRvY3VtZW50SUQ9InhtcC5kaWQ6OUM2NjlDQjM4ODBGMTFFMTg1ODlEODNERDJBRjUwQTQiPiA8eG1wTU06RGVyaXZlZEZyb20gc3RSZWY6aW5zdGFuY2VJRD0ieG1wLmlpZDo5QzY2OUNCMDg4MEYxMUUxODU4OUQ4M0REMkFGNTBBNCIgc3RSZWY6ZG9jdW1lbnRJRD0ieG1wLmRpZDo5QzY2OUNCMTg4MEYxMUUxODU4OUQ4M0REMkFGNTBBNCIvPiA8L3JkZjpEZXNjcmlwdGlvbj4gPC9yZGY6UkRGPiA8L3g6eG1wbWV0YT4gPD94cGFja2V0IGVuZD0iciI/PsQhXeAAAABfSURBVHjaYvz//z8DJYCRUgMYQAbAMBQIAvEqkBQWXI6sHqwHiwG70TTBxGaiWwjCTGgOUgJiF1J8wMRAIUA34B4Q76HUBelAfJYSA0CuMIEaRP8wGIkGMA54bgQIMACAmkXJi0hKJQAAAABJRU5ErkJggg==) no-repeat 10px center;
  text-decoration: none; }

h1 tt, h1 code {
  font-size: inherit; }

h2 tt, h2 code {
  font-size: inherit; }

h3 tt, h3 code {
  font-size: inherit; }

h4 tt, h4 code {
  font-size: inherit; }

h5 tt, h5 code {
  font-size: inherit; }

h6 tt, h6 code {
  font-size: inherit; }

h1 {
  font-size: 28px;
  color: black; }

h2 {
  font-size: 24px;
  border-bottom: 1px solid #cccccc;
  color: black; }

h3 {
  font-size: 18px; }

h4 {
  font-size: 16px; }

h5 {
  font-size: 14px; }

h6 {
  color: #777777;
  font-size: 14px; }

p, blockquote, ul, ol, dl, li, table, pre {
  margin: 15px 0; }

hr {
  background: transparent url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAYAAAAECAYAAACtBE5DAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAyJpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIiB4bWxuczp4bXBNTT0iaHR0cDovL25zLmFkb2JlLmNvbS94YXAvMS4wL21tLyIgeG1sbnM6c3RSZWY9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9zVHlwZS9SZXNvdXJjZVJlZiMiIHhtcDpDcmVhdG9yVG9vbD0iQWRvYmUgUGhvdG9zaG9wIENTNSBNYWNpbnRvc2giIHhtcE1NOkluc3RhbmNlSUQ9InhtcC5paWQ6OENDRjNBN0E2NTZBMTFFMEI3QjRBODM4NzJDMjlGNDgiIHhtcE1NOkRvY3VtZW50SUQ9InhtcC5kaWQ6OENDRjNBN0I2NTZBMTFFMEI3QjRBODM4NzJDMjlGNDgiPiA8eG1wTU06RGVyaXZlZEZyb20gc3RSZWY6aW5zdGFuY2VJRD0ieG1wLmlpZDo4Q0NGM0E3ODY1NkExMUUwQjdCNEE4Mzg3MkMyOUY0OCIgc3RSZWY6ZG9jdW1lbnRJRD0ieG1wLmRpZDo4Q0NGM0E3OTY1NkExMUUwQjdCNEE4Mzg3MkMyOUY0OCIvPiA8L3JkZjpEZXNjcmlwdGlvbj4gPC9yZGY6UkRGPiA8L3g6eG1wbWV0YT4gPD94cGFja2V0IGVuZD0iciI/PqqezsUAAAAfSURBVHjaYmRABcYwBiM2QSA4y4hNEKYDQxAEAAIMAHNGAzhkPOlYAAAAAElFTkSuQmCC) repeat-x 0 0;
  border: 0 none;
  color: #cccccc;
  height: 4px;
  padding: 0;
}

body > h2:first-child {
  margin-top: 0;
  padding-top: 0; }
body > h1:first-child {
  margin-top: 0;
  padding-top: 0; }
  body > h1:first-child + h2 {
    margin-top: 0;
    padding-top: 0; }
body > h3:first-child, body > h4:first-child, body > h5:first-child, body > h6:first-child {
  margin-top: 0;
  padding-top: 0; }

a:first-child h1, a:first-child h2, a:first-child h3, a:first-child h4, a:first-child h5, a:first-child h6 {
  margin-top: 0;
  padding-top: 0; }

h1 p, h2 p, h3 p, h4 p, h5 p, h6 p {
  margin-top: 0; }

li p.first {
  display: inline-block; }
li {
  margin: 0; }
ul, ol {
  padding-left: 30px; }

ul :first-child, ol :first-child {
  margin-top: 0; }

dl {
  padding: 0; }
  dl dt {
    font-size: 14px;
    font-weight: bold;
    font-style: italic;
    padding: 0;
    margin: 15px 0 5px; }
    dl dt:first-child {
      padding: 0; }
    dl dt > :first-child {
      margin-top: 0; }
    dl dt > :last-child {
      margin-bottom: 0; }
  dl dd {
    margin: 0 0 15px;
    padding: 0 15px; }
    dl dd > :first-child {
      margin-top: 0; }
    dl dd > :last-child {
      margin-bottom: 0; }

blockquote {
  border-left: 4px solid #dddddd;
  padding: 0 15px;
  color: #777777; }
  blockquote > :first-child {
    margin-top: 0; }
  blockquote > :last-child {
    margin-bottom: 0; }

table {
  padding: 0;border-collapse: collapse; }
  table tr {
    border-top: 1px solid #cccccc;
    background-color: white;
    margin: 0;
    padding: 0; }
    table tr:nth-child(2n) {
      background-color: #f8f8f8; }
    table tr th {
      font-weight: bold;
      border: 1px solid #cccccc;
      margin: 0;
      padding: 6px 13px; }
    table tr td {
      border: 1px solid #cccccc;
      margin: 0;
      padding: 6px 13px; }
    table tr th :first-child, table tr td :first-child {
      margin-top: 0; }
    table tr th :last-child, table tr td :last-child {
      margin-bottom: 0; }

img {
  max-width: 100%; }

span.frame {
  display: block;
  overflow: hidden; }
  span.frame > span {
    border: 1px solid #dddddd;
    display: block;
    float: left;
    overflow: hidden;
    margin: 13px 0 0;
    padding: 7px;
    width: auto; }
  span.frame span img {
    display: block;
    float: left; }
  span.frame span span {
    clear: both;
    color: #333333;
    display: block;
    padding: 5px 0 0; }
span.align-center {
  display: block;
  overflow: hidden;
  clear: both; }
  span.align-center > span {
    display: block;
    overflow: hidden;
    margin: 13px auto 0;
    text-align: center; }
  span.align-center span img {
    margin: 0 auto;
    text-align: center; }
span.align-right {
  display: block;
  overflow: hidden;
  clear: both; }
  span.align-right > span {
    display: block;
    overflow: hidden;
    margin: 13px 0 0;
    text-align: right; }
  span.align-right span img {
    margin: 0;
    text-align: right; }
span.float-left {
  display: block;
  margin-right: 13px;
  overflow: hidden;
  float: left; }
  span.float-left span {
    margin: 13px 0 0; }
span.float-right {
  display: block;
  margin-left: 13px;
  overflow: hidden;
  float: right; }
  span.float-right > span {
    display: block;
    overflow: hidden;
    margin: 13px auto 0;
    text-align: right; }

code, tt {
  margin: 0 2px;
  padding: 0 5px;
  white-space: nowrap;
  border: 1px solid #eaeaea;
  background-color: #f8f8f8;
  border-radius: 3px; }

pre code {
  margin: 0;
  padding: 0;
  white-space: pre;
  border: none;
  background: transparent; }

.highlight pre {
  background-color: #f8f8f8;
  border: 1px solid #cccccc;
  font-size: 13px;
  line-height: 19px;
  overflow: auto;
  padding: 6px 10px;
  border-radius: 3px; }

pre {
  background-color: #f8f8f8;
  border: 1px solid #cccccc;
  font-size: 13px;
  line-height: 19px;
  overflow: auto;
  padding: 6px 10px;
  border-radius: 3px; }
  pre code, pre tt {
    background-color: transparent;
    border: none; }

sup {
    font-size: 0.83em;
    vertical-align: super;
    line-height: 0;
}

kbd {
  display: inline-block;
  padding: 3px 5px;
  font-size: 11px;
  line-height: 10px;
  color: #555;
  vertical-align: middle;
  background-color: #fcfcfc;
  border: solid 1px #ccc;
  border-bottom-color: #bbb;
  border-radius: 3px;
  box-shadow: inset 0 -1px 0 #bbb
}

* {
	-webkit-print-color-adjust: exact;
}
@media screen and (min-width: 914px) {
    body {
        width: 854px;
        margin:0 auto;
    }
}
@media print {
	table, pre {
		page-break-inside: avoid;
	}
	pre {
		word-wrap: break-word;
	}
}
</style>

<style type="text/css">
/**
 * prism.js default theme for JavaScript, CSS and HTML
 * Based on dabblet (http://dabblet.com)
 * @author Lea Verou
 */

code[class*="language-"],
pre[class*="language-"] {
	color: black;
	background: none;
	text-shadow: 0 1px white;
	font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
	text-align: left;
	white-space: pre;
	word-spacing: normal;
	word-break: normal;
	word-wrap: normal;
	line-height: 1.5;

	-moz-tab-size: 4;
	-o-tab-size: 4;
	tab-size: 4;

	-webkit-hyphens: none;
	-moz-hyphens: none;
	-ms-hyphens: none;
	hyphens: none;
}

pre[class*="language-"]::-moz-selection, pre[class*="language-"] ::-moz-selection,
code[class*="language-"]::-moz-selection, code[class*="language-"] ::-moz-selection {
	text-shadow: none;
	background: #b3d4fc;
}

pre[class*="language-"]::selection, pre[class*="language-"] ::selection,
code[class*="language-"]::selection, code[class*="language-"] ::selection {
	text-shadow: none;
	background: #b3d4fc;
}

@media print {
	code[class*="language-"],
	pre[class*="language-"] {
		text-shadow: none;
	}
}

/* Code blocks */
pre[class*="language-"] {
	padding: 1em;
	margin: .5em 0;
	overflow: auto;
}

:not(pre) > code[class*="language-"],
pre[class*="language-"] {
	background: #f5f2f0;
}

/* Inline code */
:not(pre) > code[class*="language-"] {
	padding: .1em;
	border-radius: .3em;
	white-space: normal;
}

.token.comment,
.token.prolog,
.token.doctype,
.token.cdata {
	color: slategray;
}

.token.punctuation {
	color: #999;
}

.namespace {
	opacity: .7;
}

.token.property,
.token.tag,
.token.boolean,
.token.number,
.token.constant,
.token.symbol,
.token.deleted {
	color: #905;
}

.token.selector,
.token.attr-name,
.token.string,
.token.char,
.token.builtin,
.token.inserted {
	color: #690;
}

.token.operator,
.token.entity,
.token.url,
.language-css .token.string,
.style .token.string {
	color: #a67f59;
	background: hsla(0, 0%, 100%, .5);
}

.token.atrule,
.token.attr-value,
.token.keyword {
	color: #07a;
}

.token.function {
	color: #DD4A68;
}

.token.regex,
.token.important,
.token.variable {
	color: #e90;
}

.token.important,
.token.bold {
	font-weight: bold;
}
.token.italic {
	font-style: italic;
}

.token.entity {
	cursor: help;
}
</style>


</head>

<body>

<h1 id="toc_0">Doublesex Project: Step by step methods using <em>Acromyrmex echinatior</em> data</h1>

<p>Description of the pipeline used for the Doublesex project. </p>

<ul class="toc">
<li>
<a href="#toc_0">Doublesex Project: Step by step methods using <em>Acromyrmex echinatior</em> data</a>
<ul>
<li>
<a href="#toc_1">1. Getting the data</a>
<ul>
<li>
<a href="#toc_2">1. Choosing the data we want to download</a>
</li>
<li>
<a href="#toc_3">2. Downloading the SRA files</a>
</li>
<li>
<a href="#toc_4">3. Transfer to the server</a>
</li>
<li>
<a href="#toc_5">4. Conversion to fastq format</a>
</li>
<li>
<a href="#toc_6">5. Compress files</a>
</li>
</ul>
</li>
<li>
<a href="#toc_7">2. Pre-processing</a>
<ul>
<li>
<a href="#toc_8">1. Trimming the reads</a>
</li>
<li>
<a href="#toc_9">2. Quality Control: Running FASTQC</a>
</li>
</ul>
</li>
<li>
<a href="#toc_10">3. Necklace</a>
<ul>
<li>
<a href="#toc_11">1. Input for Necklace</a>
</li>
<li>
<a href="#toc_12">2. Running Necklace</a>
</li>
</ul>
</li>
<li>
<a href="#toc_13">Kallisto</a>
<ul>
<li>
<a href="#toc_14">1. Transcript-only fasta file</a>
</li>
<li>
<a href="#toc_15">2. Kallisto index file</a>
</li>
<li>
<a href="#toc_16">3. Run Kallisto</a>
</li>
<li>
<a href="#toc_17">4. Sleuth: Analysis of Kallisto&#39;s results</a>
</li>
</ul>
</li>
</ul>
</li>
</ul>


<h2 id="toc_1">1. Getting the data</h2>

<h3 id="toc_2">1. Choosing the data we want to download</h3>

<p>Data used in the project was downloaded from the <a href="https://trace.ncbi.nlm.nih.gov/Traces/sra/">SRA NCBI database</a> (public access).<br>
We need to choose the data we want to download in advance and get a list of the SRA accession numbers. Getting also the RunInfo table, which includes all the sample details, is also very helpful. For convention I have used this notation: </p>

<p><strong>Accession list:</strong><br>
Xyy_[SRA Study Number]_SRR_Acc_List.txt<br>
E.g. Aec_SRP031846_SraRunTable.txt</p>

<p><strong>RunInfo Table</strong><br>
Xyy_[SRA Study Number]_SraRunTable.txt<br>
E.g. Aec_SRP031846_SRR_Acc_List.txt</p>

<p><strong>X</strong> is the inicial of the genus name (e.g A for Acromyrmex).<br>
<strong>yy</strong> are the two fist letters of the species name (e.g. ec for echinatior)</p>

<p>Data used in the example is found in: <a href="https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP031846">Achinatior SRA data</a>.<br>
Information about the samples is downloaded from the site. Two files conatining all info and samples SRA ID&#39;s:</p>

<ol>
<li><strong>Aec_SRP031846_SraRunTable.txt</strong></li>
</ol>

<button onclick="myButton('table1')">Table</button>

<div id=table1 style="display:none;"><pre><code class="language-none">Assay_Type  BioSample   Experiment  LibrarySelection    LibrarySource   LoadDate    MBases  MBytes  Run SRA_Sample  Sample_Name caste   colony_id   source_name tissue  AvgSpotLen  BioProject  Center_Name Consent DATASTORE_filetype  DATASTORE_provider  InsertSize  Instrument  LibraryLayout   Organism    Platform    ReleaseDate SRA_Study
RNA-Seq SAMN02380854    SRX366954   cDNA    TRANSCRIPTOMIC  2015-12-15  6032    4165    SRR1015499  SRS493081   GSM1248666  gyne    Ae322   gyne heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380854    SRX366954   cDNA    TRANSCRIPTOMIC  2015-12-15  2764    1959    SRR1015500  SRS493081   GSM1248666  gyne    Ae322   gyne heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380854    SRX366954   cDNA    TRANSCRIPTOMIC  2015-12-15  3008    2105    SRR1015501  SRS493081   GSM1248666  gyne    Ae322   gyne heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380857    SRX366956   cDNA    TRANSCRIPTOMIC  2015-12-15  3793    2622    SRR1015503  SRS493083   GSM1248668  gyne    Ae356   gyne heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380857    SRX366956   cDNA    TRANSCRIPTOMIC  2015-12-15  3292    2321    SRR1015504  SRS493083   GSM1248668  gyne    Ae356   gyne heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380857    SRX366956   cDNA    TRANSCRIPTOMIC  2015-12-15  3341    2344    SRR1015505  SRS493083   GSM1248668  gyne    Ae356   gyne heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380858    SRX366958   cDNA    TRANSCRIPTOMIC  2015-12-14  3381    2368    SRR1015507  SRS493084   GSM1248670  gyne    Ae363   gyne heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380858    SRX366958   cDNA    TRANSCRIPTOMIC  2015-12-15  3171    2251    SRR1015508  SRS493084   GSM1248670  gyne    Ae363   gyne heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380858    SRX366958   cDNA    TRANSCRIPTOMIC  2015-12-15  3225    2295    SRR1015509  SRS493084   GSM1248670  gyne    Ae363   gyne heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380869    SRX366960   cDNA    TRANSCRIPTOMIC  2015-12-14  4291    2993    SRR1015511  SRS493087   GSM1248672  large worker    Ae322   large worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380869    SRX366960   cDNA    TRANSCRIPTOMIC  2015-12-15  2844    2027    SRR1015512  SRS493087   GSM1248672  large worker    Ae322   large worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380869    SRX366960   cDNA    TRANSCRIPTOMIC  2015-12-15  3093    2187    SRR1015513  SRS493087   GSM1248672  large worker    Ae322   large worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380868    SRX366962   cDNA    TRANSCRIPTOMIC  2015-12-15  3556    2475    SRR1015515  SRS493089   GSM1248674  large worker    Ae356   large worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380868    SRX366962   cDNA    TRANSCRIPTOMIC  2015-12-15  3197    2252    SRR1015516  SRS493089   GSM1248674  large worker    Ae356   large worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380868    SRX366962   cDNA    TRANSCRIPTOMIC  2015-12-15  3251    2295    SRR1015517  SRS493089   GSM1248674  large worker    Ae356   large worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380859    SRX366964   cDNA    TRANSCRIPTOMIC  2015-12-15  3597    2504    SRR1015519  SRS493091   GSM1248676  large worker    Ae363   large worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380859    SRX366964   cDNA    TRANSCRIPTOMIC  2015-12-15  3060    2160    SRR1015520  SRS493091   GSM1248676  large worker    Ae363   large worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380859    SRX366964   cDNA    TRANSCRIPTOMIC  2015-12-15  3109    2198    SRR1015521  SRS493091   GSM1248676  large worker    Ae363   large worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380867    SRX366965   cDNA    TRANSCRIPTOMIC  2015-12-14  4179    2899    SRR1015523  SRS493092   GSM1248678  small worker    Ae322   small worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380867    SRX366965   cDNA    TRANSCRIPTOMIC  2015-12-15  3667    2587    SRR1015524  SRS493092   GSM1248678  small worker    Ae322   small worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380867    SRX366965   cDNA    TRANSCRIPTOMIC  2015-12-15  3724    2624    SRR1015525  SRS493092   GSM1248678  small worker    Ae322   small worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380864    SRX366967   cDNA    TRANSCRIPTOMIC  2015-12-14  3981    2762    SRR1015527  SRS493094   GSM1248680  small worker    Ae356   small worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380864    SRX366967   cDNA    TRANSCRIPTOMIC  2015-12-15  2973    2124    SRR1015528  SRS493094   GSM1248680  small worker    Ae356   small worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380864    SRX366967   cDNA    TRANSCRIPTOMIC  2015-12-15  3563    2506    SRR1015529  SRS493094   GSM1248680  small worker    Ae356   small worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380865    SRX366969   cDNA    TRANSCRIPTOMIC  2015-12-15  3850    2669    SRR1015531  SRS493096   GSM1248682  small worker    Ae363   small worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380865    SRX366969   cDNA    TRANSCRIPTOMIC  2015-12-15  2942    2102    SRR1015532  SRS493096   GSM1248682  small worker    Ae363   small worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846
RNA-Seq SAMN02380865    SRX366969   cDNA    TRANSCRIPTOMIC  2015-12-15  3526    2485    SRR1015533  SRS493096   GSM1248682  small worker    Ae363   small worker heads  heads   180 PRJNA223531 GEO public  sra ncbi    0   Illumina HiSeq 2000 PAIRED  Acromyrmex echinatior   ILLUMINA    2015-07-22  SRP031846</code></pre></div>

<ol>
<li><strong>Aec_SRP031846_SRR_Acc_List.txt</strong></li>
</ol>

<button onclick="myButton('table2')">Table</button>
<div id=table2 style="display:none;"><pre><code class="language-none">SRR1015499
SRR1015500
SRR1015501
SRR1015503
SRR1015504
SRR1015505
SRR1015507
SRR1015508
SRR1015509
SRR1015511
SRR1015512
SRR1015513
SRR1015515
SRR1015516
SRR1015517
SRR1015519
SRR1015520
SRR1015521
SRR1015523
SRR1015524
SRR1015525
SRR1015527
SRR1015528
SRR1015529
SRR1015531
SRR1015532
SRR1015533</code></pre></div>

<h3 id="toc_3">2. Downloading the SRA files</h3>

<p>Aspera download was used to download the data files locally.</p>

<p><strong>Script: <a href="https://github.com/annafarre/Doublesex/blob/master/aspera_downloads.sh">aspera_downloads.sh</a></strong></p>

<button onclick="myButton('codeAspera')">Code</button>
<div id=codeAspera style="display:none;"><pre><code class="language-bash">#!/bin/bash
# Download a bunch of *.sra files from the NCBI SRA, using the aspera client


max_bandwidth_mbps=5000

# SRA files written line by line in a STDIN file

while read file
do 
 /Users/afarre/Applications/Aspera\ Connect.app/Contents/Resources/ascp \
 -i /Users/afarre/Applications/Aspera\ Connect.app/Contents/Resources/asperaweb_id_dsa.openssh \
 -k1 -QTr -l${max_bandwidth_mbps}m \
 anonftp@ftp-trace.ncbi.nlm.nih.gov:/sra/sra-instant/reads/ByRun/sra/${file:0:3}/${file:0:6}/${file}/${file}.sra /Users/afarre/SRA/
done &lt;&quot;$1&quot;</code></pre></div>

<p>With this script we get the SRA files to out local machine in a compressed format (file.sra) </p>

<h3 id="toc_4">3. Transfer to the server</h3>

<p>Transfer the SRA files ti the SPARTAN server using <code>rsync</code>.  </p>

<button onclick="myButton('codeRsync')">Code</button>
<div id=codeRsync style="display:none;"><pre><code class="language-none">rsync ~/SRA/* afarre@spartan.hpc.unimelb.edu.au:/data/projects/punim0356/SRA/.
afarre@spartan.hpc.unimelb.edu.au&#39;s </code></pre></div>

<h3 id="toc_5">4. Conversion to fastq format</h3>

<p>In order to be able to use the data we have to convert it to fastq format. We use the <code>fastq-dump</code> from the SRA-toolkit for that. </p>

<p>Pair end reads (PE) and single end (SE) reads requiere different commands (-r option). We can find out if a sample is PE or SE in the RunInfo Table. </p>

<p><strong>Script: <a href="https://github.com/annafarre/Doublesex/blob/master/dumpthemall.sh">dumpthemall.sh</a></strong></p>

<button onclick="myButton('codeDump')">Code</button>
<div id=codeDump style="display:none;"><pre><code class="language-bash">#!/bin/bash
# ------------------------------------------------------------------
# [Author]  Anna Farre Orteu
#           afarre@student.unimelb.edu.au
#           Fastq-dump for paired and single end sra files
# ------------------------------------------------------------------

bold=$(tput bold)
normal=$(tput sgr0)

subject=fatsq-dump
version=0.1.0
usage=&quot;
${bold}DESCRIPTION: 
${normal}Create scripts that will fastq-dump sra files for paired and single end libraries
Paired end and single end have to be run ${bold}separately
    
${bold}USAGE: 
${normal}dumpthemall [OPTION] [list of SRA files to fastq-dump] 

where:
    -h  show this help text
    -p  path to SRA files
    -r  library type: ${bold}p ${normal}(paired end) and ${bold}s ${normal}(single end)

&quot;
# --- Option processing --------------------------------------------
if [ $# == 0 ] ; then
    echo &quot;$usage&quot;
    exit 1;
fi

while getopts &quot;:p:r:h&quot; optname
  do
    case &quot;$optname&quot; in
      &quot;p&quot;)
        if [ ${OPTARG: -1} == &quot;/&quot; ] ; then
            path=&quot;$OPTARG&quot;
        else 
            path=&quot;$OPTARG/&quot;
        fi
        ;;
      &quot;r&quot;)
        if [ $OPTARG == &quot;p&quot; ] ; then
            options=&quot;--split-3 --readids --skip-technical --clip --read-filter pass --dumpbase&quot;       
            echo $options
        elif [ $OPTARG == &quot;s&quot; ] ; then
            options=&quot;--readids --skip-technical --clip --read-filter pass --dumpbase&quot;          
            echo $options
        else 
            echo &quot;Invalid argument. Specify p (paired) or s (single) end library&quot;        
        fi        
        ;;
      &quot;h&quot;)
        echo &quot;$usage&quot;
        exit 0;
        ;;
      &quot;?&quot;)
        echo &quot;Unknown option $OPTARG&quot;
        exit 0;
        ;;
      &quot;:&quot;)
        echo &quot;No argument value for option $OPTARG&quot;
        exit 0;
        ;;
      *)
        echo &quot;Unknown error while processing options&quot;
        exit 0;
        ;;
    esac
  done

if [ $OPTIND -eq 1 ]; then 
    printf &quot;\nNo options were passed
    $usage&quot; 
    exit 1;
fi

shift $(($OPTIND - 1))

param1=$1

# -----------------------------------------------------------------
#  SCRIPT LOGIC GOES HERE
# ------------------------------------------------------------
while read line
do 
    pathScript=${path}${line}_fastq-dump.sh
    cat &lt;&lt;-EOF &gt; ${pathScript}
    #!/bin/bash&quot; 
    #SBATCH -p physical&quot; 
    #SBATCH --time=03:00:00&quot; 
    #SBATCH --nodes=1&quot; 
    #SBATCH --ntasks=1&quot; 
    #SBATCH --cpus-per-task=1&quot; 
    #SBATCH --job-name=fastq-dump&quot; 
    #SBATCH --mem-per-cpu=50000&quot; 
    #SBATCH --mail-type=ALL&quot; 
    #SBATCH --mail-user=afarre@student.unimelb.edu.au&quot; 
    #SBATCH --out=slurm_%j.out&quot; 
    #SBATCH --err=slurm_%j.err&quot; 

    fastq-dump $options ${path}${line}.sra 
    EOF
    
done &lt;&quot;$1&quot;</code></pre></div>

<p><strong>Commands:</strong></p>

<div><pre><code class="language-bash">[afarre@spartan SRA]$ pwd
/data/projects/punim0356/SRA
[afarre@spartan SRA]$ bash ~/scripts/dumpthemall.sh -p . -r p ../data_info/Aec_SRP031846_SRR_Acc_List.txt </code></pre></div>

<h3 id="toc_6">5. Compress files</h3>

<p>In order to save space in the server, we compress the <code>.fastq</code> files to <code>.fastq.gz.</code> </p>

<p><strong>Script: gzip.sh</strong></p>

<button onclick="myButton('codeGzip')">Code</button>
<div id=codeGzip style="display:none;"><pre><code class="language-bash">#!/bin/bash

# Gunzip all files listed
# Input through STDIN 
# Use wildcards to zip multiple files
# Creates a script per file

for filename in `ls $@`
do

pathScript=$(pwd)/gzip_${filename}.sh

cat &lt;&lt;-EOF &gt; ${pathScript}
#!/bin/bash
 
#SBATCH -p physical
#SBATCH --time=9-24
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=1
#SBATCH --job-name=gzip_${filename}
#SBATCH --mem=25GB
#SBATCH --mail-type=ALL
#SBATCH --mail-user=afarre@student.unimelb.edu.au
#SBATCH --out=slurm_%j.out
#SBATCH --err=slurm_%j.err
 
gzip ${filename}

EOF
done </code></pre></div>

<div><pre><code class="language-bash">[afarre@spartan SRA]$ pwd
/data/projects/punim0356/SRA
[afarre@spartan SRA]$ 
bash ~/scripts/gzip.sh *fastq</code></pre></div>

<h2 id="toc_7">2. Pre-processing</h2>

<p>Once the data is in fastq format we need some steps of pre-procession. </p>

<h3 id="toc_8">1. Trimming the reads</h3>

<p>We use TrimGalore to trim adapters and low quality parts of the reads.</p>

<p>Pair end reads (PE) and single end (SE) reads requiere different commands/scripts. We can find out if a sample is PE or SE in the RunInfo Table. </p>

<p>The script can run on compressed (<code>.gz</code>) data. </p>

<p><strong>Script for PE: <a href="https://github.com/annafarre/Doublesex/blob/master/trimGalore.sh">trimGalore.sh</a></strong></p>

<button onclick="myButton('codeTrim')">Code</button>
<div id=codeTrim style="display:none;"><pre><code class="language-bash">#!/bin/bash
# ------------------------------------------------------------------
# [Author]  Anna Farre Orteu, 2017
#           afarre@student.unimelb.edu.au
#           Create scripts to run TrimGalore!
# ------------------------------------------------------------------

bold=$(tput bold)
normal=$(tput sgr0)

subject=TrimGalore!
usage=&quot;
${bold}DESCRIPTION: 
${normal}Create scripts that will run TrimGalore!
Trims adapters, low quality read ends and runs FastQC
Paired end data ONLY
gz compressed data
extention must be: 
    *_pass_1.fastq.gz 
    *_pass_2.fastq.gz
    
trimGalore.sh [options] list_SRA.txt

where: 
    -o output path
    -i input path

${bold}REQUIEREMENTS
${normal}cutadapt
fastqc
&quot;

# --- Option processing --------------------------------------------
if [ $# == 0 ] ; then
    echo &quot;$usage&quot;
    exit 1;
fi
while getopts &quot;:o:i:h&quot; optname
  do
    case &quot;$optname&quot; in
      &quot;i&quot;)
        if [[ $OPTARG == /* ]] ; then
            if [ ${OPTARG: -1} == &quot;/&quot; ] ; then
                inputPath=&quot;$OPTARG&quot;
            else 
                inputPath=&quot;$OPTARG/&quot;
            fi
       
        else 
            if [ ${OPTARG: -1} == &quot;/&quot; ] ; then
                inputPath=&quot;$(pwd)/$OPTARG&quot;
            else 
                inputPath=&quot;$(pwd)/$OPTARG/&quot;
            fi
        fi
        ;;
      &quot;o&quot;) 
        mkdir -p $OPTARG
        
        if [[ $OPTARG == /* ]] ; then
            if [ ${OPTARG: -1} == &quot;/&quot; ] ; then
                outputPath=&quot;$OPTARG&quot;
            else 
                outputPath=&quot;$OPTARG/&quot;
            fi
       
        else 
            if [ ${OPTARG: -1} == &quot;/&quot; ] ; then
                outputPath=&quot;$(pwd)/$OPTARG&quot;
            else 
                outputPath=&quot;$(pwd)/$OPTARG/&quot;
            fi
        fi
        ;;      
      &quot;h&quot;)
        echo &quot;$usage&quot;
        exit 0;
        ;;
      &quot;?&quot;)
        echo &quot;Unknown option $OPTARG&quot;
        exit 0;
        ;;
      &quot;:&quot;)
        echo &quot;No argument value for option $OPTARG&quot;
        exit 0;
        ;;
      *)
        echo &quot;Unknown error while processing options&quot;
        exit 0;
        ;;
    esac
  done
  
if [ $OPTIND -eq 1 ]; then 
    printf &quot;\nNo options were passed
    $usage&quot; 
    exit 1;
fi

shift $(($OPTIND - 1))

# -----------------------------------------------------------------
#  SCRIPT LOGIC GOES HERE
# -----------------------------------------------------------------


list=$@

while read filename 
do

pathScript=${outputPath}trimGalore_${filename}.sh

cat &lt;&lt;-EOF &gt; ${pathScript}
#!/bin/bash

#run TrimGalore on raw data 

#SBATCH -p physical
#SBATCH --time=03:00:00
#SBATCH --job-name=trimGalore_${filename}
#SBATCH --mem=25GB
#SBATCH --mail-type=ALL
#SBATCH --mail-user=afarre@student.unimelb.edu.au
#SBATCH --out=slurm_trimGalore_${filename}_%j.out
#SBATCH --err=slurm_trimGalore_${filename}_%j.err


#path to list of files to use

module load Python
module load fastqc

/home/afarre/TrimGalore-0.4.5/trim_galore --fastqc --gzip --output_dir ${outputPath} --paired ${inputPath}${filename}_pass_1.fastq.gz ${inputPath}${filename}_pass_2.fastq.gz
EOF
done &lt;&quot;${list}&quot;</code></pre></div>

<h3 id="toc_9">2. Quality Control: Running FASTQC</h3>

<p>To check the quality of the reads we use FASTQC. We can run it on the raw reads or in the trimmed ones.  </p>

<p>Fatqc creates multible html files with results of the quality contol tests. </p>

<p><strong>Script: <a href="https://github.com/annafarre/Doublesex/blob/master/fastqcthemall.sh">fastqcthemall.sh</a></strong> </p>

<button onclick="myButton('codeFastqc')">Code</button>
<div id=codeFastqc style="display:none;"><pre><code class="language-bash">#!/bin/bash
# ------------------------------------------------------------------
# [Author]  Anna Farre Orteu
#           afarre@student.unimelb.edu.au
#           Fastqc for paired and single end fastq files
# ------------------------------------------------------------------

bold=$(tput bold)
normal=$(tput sgr0)

subject=fatsq-dump
version=0.1.0
usage=&quot;
${bold}DESCRIPTION: 
${normal}Run fastqc in paired and single end fastq files
    
${bold}USAGE: 
${normal}fastqc.sh [OPTION] [list of fastq files] 

where:
    -h  show this help text
    -o  path where output has to be saved
    
&quot;
# --- Option processing --------------------------------------------
if [ $# == 0 ] ; then
    echo &quot;$usage&quot;
    exit 1;
fi

while getopts &quot;:o:h&quot; optname
  do
    case &quot;$optname&quot; in
      &quot;o&quot;)
        mkdir -p $OPTARG
        
        if [[ $OPTARG == /* ]] ; then
            if [ ${OPTARG: -1} == &quot;/&quot; ] ; then
                outputPath=&quot;$OPTARG&quot;
            else 
                outputPath=&quot;$OPTARG/&quot;
            fi
       
        else 
            if [ ${OPTARG: -1} == &quot;/&quot; ] ; then
                outputPath=&quot;$(pwd)/$OPTARG&quot;
            else 
                outputPath=&quot;$(pwd)/$OPTARG/&quot;
            fi
        fi
        ;;
      &quot;h&quot;)
        echo &quot;$usage&quot;
        exit 0;
        ;;
      &quot;?&quot;)
        echo &quot;Unknown option $OPTARG&quot;
        exit 0;
        ;;
      &quot;:&quot;)
        echo &quot;No argument value for option $OPTARG&quot;
        exit 0;
        ;;
      *)
        echo &quot;Unknown error while processing options&quot;
        exit 0;
        ;;
    esac
  done

if [ $OPTIND -eq 1 ]; then 
    printf &quot;\nNo options were passed
    $usage&quot; 
    exit 1;
fi

shift $(($OPTIND - 1))

param1=$@

# -----------------------------------------------------------------
#  SCRIPT LOGIC GOES HERE
# ------------------------------------------------------------

for filename in `ls ${param1}`
do

pathScript=fastqc_${filename}.sh
cat &lt;&lt;-EOF &gt; ${pathScript}
#!/bin/bash

#SBATCH -p cloud
#SBATCH --time=09:00:00 
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=8
#SBATCH --job-name=fastqc_${filename}
#SBATCH --mem-per-cpu=4000
#SBATCH --mail-type=ALL
#SBATCH --mail-user=afarre@student.unimelb.edu.au
#SBATCH --out=slurm_%j.out
#SBATCH --err=slurm_%j.err 

module load fastqc

fastqc -o ${outputPath} -t 8 ${filename}
EOF
done</code></pre></div>

<h2 id="toc_10">3. Necklace</h2>

<p><a href="https://github.com/Oshlack/necklace/wiki">Necklace</a> is a pipeline for RNA-seq analysis developed by the <a href="http://oshlacklab.com/">Oschlack Lab</a>. It was made for RNA-Seq analyses involving species with incomplete genome or annotations. ie. most organisms other than human, mouse, drosophila etc.</p>

<h3 id="toc_11">1. Input for Necklace</h3>

<p>Necklace takes as input RNA-seq files. However read names have to follow a very specific pattern: </p>

<p><strong>DRRXXX_YYYY/Z</strong>   </p>

<p><strong>DRRXXX</strong> SRA sample name/number<br>
<strong>YYYY</strong> read number<br>
<strong>Z</strong> read pair number - 1 or 2  </p>

<p><strong>Script: <a href="https://github.com/annafarre/Doublesex/blob/master/readRename.sh">readRename.sh</a></strong></p>

<button onclick="myButton('codeReadrename')">Code</button>
<div id=codeReadrename style="display:none;"><pre><code class="language-bash">#!/bin/bash

#Rename reads in fastq.gz files coming from SRA 
#to be used in Necklace 

#Example: 
    #Old name: DRRXXX.YYYY.Z 
    #New name: DRRXXX_YYYY/Z 
    
    #DRRXXX SRA sample name/number
    #YYYY read number
    #Z read pair number - 1 or 2


for filename in `ls $@`
do
cat &lt;&lt;-EOF &gt; ${filename}_readRename.sh
#!/bin/bash

#SBATCH -p physical
#SBATCH --time=01:00:00
#SBATCH --job-name=readconvert
#SBATCH --mem=10GB
#SBATCH --out=slurm_%j.out  
#SBATCH --err=slurm_%j.err  
#SBATCH --mail-user=afarre@student.unimelb.edu.au
#SBATCH --mail-type=ALL


gunzip -c ${filename} | sed &#39;s/\(^@.\)*\.\(.*\)\.\([1-2]\)/\1_\2\/\3/g&#39; | gzip &gt; necklaceInput/${filename}
EOF
done</code></pre></div>

<h3 id="toc_12">2. Running Necklace</h3>

<p>Necklace gets the information about the input from a <code>.config</code> file.   </p>

<p><strong>Script: Vem_necklace.config</strong> </p>

<button onclick="myButton('codeConfig')">Code</button>
<div id=codeConfig style="display:none;"><pre><code class="language-bash">// sequencing data
reads_R1=&quot;/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030152_pass_1_val_1.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030153_pass_1_val_1.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030154_pass_1_val_1.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030155_pass_1_val_1.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030156_pass_1_val_1.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030157_pass_1_val_1.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030158_pass_1_val_1.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030159_pass_1_val_1.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030160_pass_1_val_1.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030161_pass_1_val_1.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030162_pass_1_val_1.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030163_pass_1_val_1.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030164_pass_1_val_1.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030165_pass_1_val_1.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030166_pass_1_val_1.fastq.gz&quot; 
reads_R2=&quot;/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030152_pass_2_val_2.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030153_pass_2_val_2.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030154_pass_2_val_2.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030155_pass_2_val_2.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030156_pass_2_val_2.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030157_pass_2_val_2.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030158_pass_2_val_2.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030159_pass_2_val_2.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030160_pass_2_val_2.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030161_pass_2_val_2.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030162_pass_2_val_2.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030163_pass_2_val_2.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030164_pass_2_val_2.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030165_pass_2_val_2.fastq.gz,/data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/DRR030166_pass_2_val_2.fastq.gz&quot;

//The reference genome and its annotation
annotation=&quot;/data/projects/punim0304/Vemeryi/genomes/GCF_000949405.1_V.emery_V1.0_genomic.gff&quot;
genome=&quot;/data/projects/punim0304/Vemeryi/genomes/GCF_000949405.1_V.emery_V1.0_genomic.fna&quot;

//The genome and annotation of a related species
annotation_related_species=&quot;/home/afarre/original_genomes/Drosophila_melanogaster.BDGP6.37.gtf&quot;
genome_related_species=&quot;/home/afarre/original_genomes/Drosophila_melanogaster.BDGP6.dna.toplevel.fa&quot;</code></pre></div>

<p>It can be created with the following <strong>commands</strong>: </p>

<button onclick="myButton('codeNecklaceCommands')">Code</button>
<div id=codeNecklaceCommands style="display:none;"><pre><code class="language-none">#IMPORTANT: no spaces between samples. Coma separated without spaces.
#ls -m -&gt; will produce spaces. Remove them before running

echo -e //Sequencing data&#39;\n&#39;reads_R1=\&quot;$(ls -m /data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/*1.fastq.gz)\&quot; &#39;\n&#39;reads_R2=\&quot;$(ls -m /data/projects/punim0304/Vemeryi/SRA/trimGalore/necklaceInput/*2.fastq.gz)\&quot; &gt; Vem_necklace.config 
echo -e &#39;\n&#39;//The reference genome and its annotation&#39;\n&#39;annotation=\&quot;$(ls /data/projects/punim0304/Vemeryi/genome/*gff)\&quot; &gt;&gt;Vem_necklace.config 
echo -e genome=\&quot;$(ls /data/projects/punim0304/Vemeryi/genome/*fna)\&quot; &gt;&gt;Vem_necklace.config 

echo -e  &#39;\n&#39;//The genome and annotation of a related species &gt;&gt;Vem_necklace.config 
echo -e  annotation_related_species=&quot;/home/afarre/original_genomes/Drosophila_melanogaster.BDGP6.37.gtf&quot; &gt;&gt;Vem_necklace.config 
echo -e  genome_related_species=&quot;/home/afarre/original_genomes/Drosophila_melanogaster.BDGP6.dna.toplevel.fa&quot; &gt;&gt;Vem_necklace.config 

echo -e  &#39;\n&#39;//The genome and annotation of a related species &gt;&gt;Vem_necklace.config  
echo -e  annotation_related_species=&quot;/home/afarre/original_genomes/Drosophila_melanogaster.BDGP6.37.gtf&quot; &gt;&gt;Vem_necklace.config 
echo -e  genome_related_species=&quot;/home/afarre/original_genomes/Drosophila_melanogaster.BDGP6.dna.toplevel.fa&quot; &gt;&gt;Vem_necklace.config  </code></pre></div>

<p>Then Necklace can be run with <strong>run_necklace.sh</strong>:</p>

<button onclick="myButton('codeRunNecklace')">Code</button>
<div id=codeRunNecklace style="display:none;"><pre><code class="language-bash">#!/bin/bash

#SBATCH -p physical
#SBATCH --time=3-12
#SBATCH --job-name=necklace
#SBATCH --cpus-per-task=10
#SBATCH --mem=100GB
#SBATCH --mail-type=ALL
#SBATCH --mail-user=afarre@student.unimelb.edu.au
#SBATCH --out=slurm_%j.out  
#SBATCH --err=slurm_%j.err 

module load Java

cd /home/afarre/Vem_necklace/

MAX_JAVA_MEM=2g /home/afarre/.local/necklace-necklace_v0.9/tools/bin/bpipe run /home/afarre/.local/necklace-necklace_v0.9/necklace.groovy /home/afarre/Vem_necklace/Vem_necklace.config </code></pre></div>

<h2 id="toc_13">Kallisto</h2>

<h3 id="toc_14">1. Transcript-only fasta file</h3>

<p>To run Kallisto first we have to create an index file that the program will use to do the pseudo alignment. This index file conatins <strong>only transcripts</strong>. Then the fist step is to create a fasta file with the only the transcript sequences. </p>

<p><strong>Commands</strong></p>

<div><pre><code class="language-bash">[afarre@spartan Vem_necklace]$ pwd
/home/afarre/Vem_necklace/
[afarre@spartan Vem_necklace]$ module load Cufflinks
[afarre@spartan Vem_necklace]$ gffread -w GCF_000949405.1_V.emery_V1.0_genomic_kallisto_genome_merged_ST.fasta -g ~/Vemeryi/genomes/GCF_000949405.1_V.emery_V1.0_genomic.fna</code></pre></div>

<h3 id="toc_15">2. Kallisto index file</h3>

<p>Using the fasta file conataining only transcript sequences, create an index file.</p>

<p><strong>Script: <a href="https://github.com/annafarre/Doublesex/blob/master/kallisto_index.sh">kallisto_index.sh</a></strong></p>

<button onclick="myButton('codeIndexKallisto')">Code</button>
<div id=codeIndexKallisto style="display:none;"><pre><code class="language-bash">#!/bin/bash

#create an index for kallisto
#input: $1 index name $2 fasta file (transcripts)


#SBATCH -p physical
#SBATCH --time=01:00:00
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=1
#SBATCH --job-name=kallisto_idx
#SBATCH --mem=20GB
#SBATCH --mail-type=ALL
#SBATCH --out=slurm_%j.out
#SBATCH --err=slurm_%j.err
#SBATCH --mail-user=afarre@student.unimelb.edu.au


kallisto index --index=$1 $2</code></pre></div>

<h3 id="toc_16">3. Run Kallisto</h3>

<p>Paired-end and single-end samples require different scripts/commands.</p>

<p><strong>Script: <a href="https://github.com/annafarre/Doublesex/blob/master/kallisto_pe.sh">kallisto_pe.sh</a></strong></p>

<button onclick="myButton('codeKallisto')">Code</button>
<div id=codeKallisto style="display:none;"><pre><code class="language-bash">#!/bin/bash
# ------------------------------------------------------------------
# [Author]  Anna Farre Orteu, 2017
#           afarre@student.unimelb.edu.au
#           Create scripts to run kallisto
# ------------------------------------------------------------------

bold=$(tput bold)
normal=$(tput sgr0)

subject=kallisto
usage=&quot;
${bold}DESCRIPTION: 
${normal}Run kallisto for paired end data
use v0.43.0 when multithreading, v0.43.1 crashes

Only for ${bold}paired-end ${normal}samples

kallisto.sh [options] list.txt

where: 
    -o output path
    -i kallisto index file
    -f input path (where fastq files are)
    -r extension for R1 (read 1) e.g _pass_1_val_1.fq.gz
    -l extension for R2 (read 2) e.g _pass_1_val_1.fq.gz
    
    list.txt contains all SRA sample IDs

&quot;

# --- Option processing --------------------------------------------
if [ $# == 0 ] ; then
    echo &quot;$usage&quot;
    exit 1;
fi
while getopts &quot;:o:i:f:r:l:h&quot; optname
  do
    case &quot;$optname&quot; in
      &quot;i&quot;)
        kallisto_index=&quot;$(pwd)/$OPTARG&quot;
        ;;
      &quot;o&quot;) 
        mkdir -p $OPTARG
        
        if [[ $OPTARG == /* ]] ; then
            if [ ${OPTARG: -1} == &quot;/&quot; ] ; then
                outputPath=&quot;$OPTARG&quot;
            else 
                outputPath=&quot;$OPTARG/&quot;
            fi
       
        else 
            if [ ${OPTARG: -1} == &quot;/&quot; ] ; then
                outputPath=&quot;$(pwd)/$OPTARG&quot;
            else 
                outputPath=&quot;$(pwd)/$OPTARG/&quot;
            fi
        fi
        ;;      
      &quot;f&quot;)          
        if [[ $OPTARG == /* ]] ; then
            if [ ${OPTARG: -1} == &quot;/&quot; ] ; then
                inputPath=&quot;$OPTARG&quot;
            else 
                inputPath=&quot;$OPTARG/&quot;
            fi
       
        else 
            if [ ${OPTARG: -1} == &quot;/&quot; ] ; then
                inputPath=&quot;$(pwd)/$OPTARG&quot;
            else 
                inputPath=&quot;$(pwd)/$OPTARG/&quot;
            fi
        fi
        ;;   
      &quot;r&quot;)
        read1Name=&quot;$OPTARG&quot;
        ;;
      &quot;l&quot;)
        read2Name=&quot;$OPTARG&quot;
        ;;   
      &quot;h&quot;)
        echo &quot;$usage&quot;
        exit 0;
        ;;
      &quot;?&quot;)
        echo &quot;Unknown option $OPTARG&quot;
        exit 0;
        ;;
      &quot;:&quot;)
        echo &quot;No argument value for option $OPTARG&quot;
        exit 0;
        ;;
      *)
        echo &quot;Unknown error while processing options&quot;
        exit 0;
        ;;
    esac
  done
  
if [ $OPTIND -eq 1 ]; then 
    printf &quot;\nNo options were passed
    $usage&quot; 
    exit 1;
fi

shift $(($OPTIND - 1))

# -----------------------------------------------------------------
#  SCRIPT LOGIC GOES HERE
# -----------------------------------------------------------------


list=&quot;$@&quot;
while read filename 
do

mkdir -p ${outputPath}${subject}_${filename}/
pathScript=${outputPath}${subject}_${filename}/${subject}_${filename}.sh

cat &lt;&lt;-EOF &gt; ${pathScript}
#!/bin/bash


#SBATCH -p physical
#SBATCH --time=03:00:00
#SBATCH --job-name=${subject}_${filename}
#SBATCH --cpus-per-task=8
#SBATCH --mem=40GB
#SBATCH --mail-type=ALL
#SBATCH --mail-user=afarre@student.unimelb.edu.au
#SBATCH --out=slurm_${subject}_${filename}_%j.out
#SBATCH --err=slurm_${subject}_${filename}_%j.err


/home/afarre/.local/kallisto_linux-v0.43.0/kallisto quant --bootstrap-samples 100 --index=${kallisto_index} --output-dir=${outputPath}${subject}_${filename}/ --threads=8 ${inputPath}${filename}${read1Name} ${inputPath}${filename}${read2Name}

EOF
done&lt;&quot;${list}&quot;</code></pre></div>

<h3 id="toc_17">4. Sleuth: Analysis of Kallisto&#39;s results</h3>



<script type="text/javascript">
var _self="undefined"!=typeof window?window:"undefined"!=typeof WorkerGlobalScope&&self instanceof WorkerGlobalScope?self:{},Prism=function(){var e=/\blang(?:uage)?-(\w+)\b/i,t=0,n=_self.Prism={util:{encode:function(e){return e instanceof a?new a(e.type,n.util.encode(e.content),e.alias):"Array"===n.util.type(e)?e.map(n.util.encode):e.replace(/&/g,"&amp;").replace(/</g,"&lt;").replace(/\u00a0/g," ")},type:function(e){return Object.prototype.toString.call(e).match(/\[object (\w+)\]/)[1]},objId:function(e){return e.__id||Object.defineProperty(e,"__id",{value:++t}),e.__id},clone:function(e){var t=n.util.type(e);switch(t){case"Object":var a={};for(var r in e)e.hasOwnProperty(r)&&(a[r]=n.util.clone(e[r]));return a;case"Array":return e.map&&e.map(function(e){return n.util.clone(e)})}return e}},languages:{extend:function(e,t){var a=n.util.clone(n.languages[e]);for(var r in t)a[r]=t[r];return a},insertBefore:function(e,t,a,r){r=r||n.languages;var l=r[e];if(2==arguments.length){a=arguments[1];for(var i in a)a.hasOwnProperty(i)&&(l[i]=a[i]);return l}var o={};for(var s in l)if(l.hasOwnProperty(s)){if(s==t)for(var i in a)a.hasOwnProperty(i)&&(o[i]=a[i]);o[s]=l[s]}return n.languages.DFS(n.languages,function(t,n){n===r[e]&&t!=e&&(this[t]=o)}),r[e]=o},DFS:function(e,t,a,r){r=r||{};for(var l in e)e.hasOwnProperty(l)&&(t.call(e,l,e[l],a||l),"Object"!==n.util.type(e[l])||r[n.util.objId(e[l])]?"Array"!==n.util.type(e[l])||r[n.util.objId(e[l])]||(r[n.util.objId(e[l])]=!0,n.languages.DFS(e[l],t,l,r)):(r[n.util.objId(e[l])]=!0,n.languages.DFS(e[l],t,null,r)))}},plugins:{},highlightAll:function(e,t){var a={callback:t,selector:'code[class*="language-"], [class*="language-"] code, code[class*="lang-"], [class*="lang-"] code'};n.hooks.run("before-highlightall",a);for(var r,l=a.elements||document.querySelectorAll(a.selector),i=0;r=l[i++];)n.highlightElement(r,e===!0,a.callback)},highlightElement:function(t,a,r){for(var l,i,o=t;o&&!e.test(o.className);)o=o.parentNode;o&&(l=(o.className.match(e)||[,""])[1],i=n.languages[l]),t.className=t.className.replace(e,"").replace(/\s+/g," ")+" language-"+l,o=t.parentNode,/pre/i.test(o.nodeName)&&(o.className=o.className.replace(e,"").replace(/\s+/g," ")+" language-"+l);var s=t.textContent,u={element:t,language:l,grammar:i,code:s};if(!s||!i)return n.hooks.run("complete",u),void 0;if(n.hooks.run("before-highlight",u),a&&_self.Worker){var c=new Worker(n.filename);c.onmessage=function(e){u.highlightedCode=e.data,n.hooks.run("before-insert",u),u.element.innerHTML=u.highlightedCode,r&&r.call(u.element),n.hooks.run("after-highlight",u),n.hooks.run("complete",u)},c.postMessage(JSON.stringify({language:u.language,code:u.code,immediateClose:!0}))}else u.highlightedCode=n.highlight(u.code,u.grammar,u.language),n.hooks.run("before-insert",u),u.element.innerHTML=u.highlightedCode,r&&r.call(t),n.hooks.run("after-highlight",u),n.hooks.run("complete",u)},highlight:function(e,t,r){var l=n.tokenize(e,t);return a.stringify(n.util.encode(l),r)},tokenize:function(e,t){var a=n.Token,r=[e],l=t.rest;if(l){for(var i in l)t[i]=l[i];delete t.rest}e:for(var i in t)if(t.hasOwnProperty(i)&&t[i]){var o=t[i];o="Array"===n.util.type(o)?o:[o];for(var s=0;s<o.length;++s){var u=o[s],c=u.inside,g=!!u.lookbehind,h=!!u.greedy,f=0,d=u.alias;u=u.pattern||u;for(var p=0;p<r.length;p++){var m=r[p];if(r.length>e.length)break e;if(!(m instanceof a)){u.lastIndex=0;var y=u.exec(m),v=1;if(!y&&h&&p!=r.length-1){var b=r[p+1].matchedStr||r[p+1],k=m+b;if(p<r.length-2&&(k+=r[p+2].matchedStr||r[p+2]),u.lastIndex=0,y=u.exec(k),!y)continue;var w=y.index+(g?y[1].length:0);if(w>=m.length)continue;var _=y.index+y[0].length,P=m.length+b.length;if(v=3,P>=_){if(r[p+1].greedy)continue;v=2,k=k.slice(0,P)}m=k}if(y){g&&(f=y[1].length);var w=y.index+f,y=y[0].slice(f),_=w+y.length,S=m.slice(0,w),O=m.slice(_),j=[p,v];S&&j.push(S);var A=new a(i,c?n.tokenize(y,c):y,d,y,h);j.push(A),O&&j.push(O),Array.prototype.splice.apply(r,j)}}}}}return r},hooks:{all:{},add:function(e,t){var a=n.hooks.all;a[e]=a[e]||[],a[e].push(t)},run:function(e,t){var a=n.hooks.all[e];if(a&&a.length)for(var r,l=0;r=a[l++];)r(t)}}},a=n.Token=function(e,t,n,a,r){this.type=e,this.content=t,this.alias=n,this.matchedStr=a||null,this.greedy=!!r};if(a.stringify=function(e,t,r){if("string"==typeof e)return e;if("Array"===n.util.type(e))return e.map(function(n){return a.stringify(n,t,e)}).join("");var l={type:e.type,content:a.stringify(e.content,t,r),tag:"span",classes:["token",e.type],attributes:{},language:t,parent:r};if("comment"==l.type&&(l.attributes.spellcheck="true"),e.alias){var i="Array"===n.util.type(e.alias)?e.alias:[e.alias];Array.prototype.push.apply(l.classes,i)}n.hooks.run("wrap",l);var o="";for(var s in l.attributes)o+=(o?" ":"")+s+'="'+(l.attributes[s]||"")+'"';return"<"+l.tag+' class="'+l.classes.join(" ")+'" '+o+">"+l.content+"</"+l.tag+">"},!_self.document)return _self.addEventListener?(_self.addEventListener("message",function(e){var t=JSON.parse(e.data),a=t.language,r=t.code,l=t.immediateClose;_self.postMessage(n.highlight(r,n.languages[a],a)),l&&_self.close()},!1),_self.Prism):_self.Prism;var r=document.currentScript||[].slice.call(document.getElementsByTagName("script")).pop();return r&&(n.filename=r.src,document.addEventListener&&!r.hasAttribute("data-manual")&&document.addEventListener("DOMContentLoaded",n.highlightAll)),_self.Prism}();"undefined"!=typeof module&&module.exports&&(module.exports=Prism),"undefined"!=typeof global&&(global.Prism=Prism);
</script>

<script type="text/javascript">
!function(e){var t={variable:[{pattern:/\$?\(\([\w\W]+?\)\)/,inside:{variable:[{pattern:/(^\$\(\([\w\W]+)\)\)/,lookbehind:!0},/^\$\(\(/],number:/\b-?(?:0x[\dA-Fa-f]+|\d*\.?\d+(?:[Ee]-?\d+)?)\b/,operator:/--?|-=|\+\+?|\+=|!=?|~|\*\*?|\*=|\/=?|%=?|<<=?|>>=?|<=?|>=?|==?|&&?|&=|\^=?|\|\|?|\|=|\?|:/,punctuation:/\(\(?|\)\)?|,|;/}},{pattern:/\$\([^)]+\)|`[^`]+`/,inside:{variable:/^\$\(|^`|\)$|`$/}},/\$(?:[a-z0-9_#\?\*!@]+|\{[^}]+\})/i]};e.languages.bash={shebang:{pattern:/^#!\s*\/bin\/bash|^#!\s*\/bin\/sh/,alias:"important"},comment:{pattern:/(^|[^"{\\])#.*/,lookbehind:!0},string:[{pattern:/((?:^|[^<])<<\s*)(?:"|')?(\w+?)(?:"|')?\s*\r?\n(?:[\s\S])*?\r?\n\2/g,lookbehind:!0,inside:t},{pattern:/(["'])(?:\\\\|\\?[^\\])*?\1/g,inside:t}],variable:t.variable,"function":{pattern:/(^|\s|;|\||&)(?:alias|apropos|apt-get|aptitude|aspell|awk|basename|bash|bc|bg|builtin|bzip2|cal|cat|cd|cfdisk|chgrp|chmod|chown|chroot|chkconfig|cksum|clear|cmp|comm|command|cp|cron|crontab|csplit|cut|date|dc|dd|ddrescue|df|diff|diff3|dig|dir|dircolors|dirname|dirs|dmesg|du|egrep|eject|enable|env|ethtool|eval|exec|expand|expect|export|expr|fdformat|fdisk|fg|fgrep|file|find|fmt|fold|format|free|fsck|ftp|fuser|gawk|getopts|git|grep|groupadd|groupdel|groupmod|groups|gzip|hash|head|help|hg|history|hostname|htop|iconv|id|ifconfig|ifdown|ifup|import|install|jobs|join|kill|killall|less|link|ln|locate|logname|logout|look|lpc|lpr|lprint|lprintd|lprintq|lprm|ls|lsof|make|man|mkdir|mkfifo|mkisofs|mknod|more|most|mount|mtools|mtr|mv|mmv|nano|netstat|nice|nl|nohup|notify-send|nslookup|open|op|passwd|paste|pathchk|ping|pkill|popd|pr|printcap|printenv|printf|ps|pushd|pv|pwd|quota|quotacheck|quotactl|ram|rar|rcp|read|readarray|readonly|reboot|rename|renice|remsync|rev|rm|rmdir|rsync|screen|scp|sdiff|sed|seq|service|sftp|shift|shopt|shutdown|sleep|slocate|sort|source|split|ssh|stat|strace|su|sudo|sum|suspend|sync|tail|tar|tee|test|time|timeout|times|touch|top|traceroute|trap|tr|tsort|tty|type|ulimit|umask|umount|unalias|uname|unexpand|uniq|units|unrar|unshar|uptime|useradd|userdel|usermod|users|uuencode|uudecode|v|vdir|vi|vmstat|wait|watch|wc|wget|whereis|which|who|whoami|write|xargs|xdg-open|yes|zip)(?=$|\s|;|\||&)/,lookbehind:!0},keyword:{pattern:/(^|\s|;|\||&)(?:let|:|\.|if|then|else|elif|fi|for|break|continue|while|in|case|function|select|do|done|until|echo|exit|return|set|declare)(?=$|\s|;|\||&)/,lookbehind:!0},"boolean":{pattern:/(^|\s|;|\||&)(?:true|false)(?=$|\s|;|\||&)/,lookbehind:!0},operator:/&&?|\|\|?|==?|!=?|<<<?|>>|<=?|>=?|=~/,punctuation:/\$?\(\(?|\)\)?|\.\.|[{}[\];]/};var a=t.variable[1].inside;a["function"]=e.languages.bash["function"],a.keyword=e.languages.bash.keyword,a.boolean=e.languages.bash.boolean,a.operator=e.languages.bash.operator,a.punctuation=e.languages.bash.punctuation}(Prism);
function myButton(id) {
	var e = document.getElementById(id);
    e.style.display = (e.style.display == 'block') ? 'none' : 'block';
}
</script>


</body>

</html>
