---
title: "Mph_necklace_DEXSeq"
author: "Anna Orteu"
date: "12/06/2018"
output: 
  html_document:
    toc: true # show a table of contents
    toc_float: true # make table of contents float at the side of the page
    depth: 4  # upto 4 depths of headings (specified by #, ##, ### and ####)
    number_sections: false  # set to true if you want to number sections at each table header
    theme: flatly
    code_folding: hide # Add some awesome buttons to show/hide the code
    df_print: paged
---

Script to analyse necklace output using DEXSeq. 
Made for non-model insects - find orthology to _Apis mellifera_ genome. 

Input requited: 

* relST_genomeST.psl file from necklace

* _Apis mellifera_ gtf file GCF_000002195.4_Amel_4.5_genomic.gff

* block.counts file from necklace

* X_SraRunTable.txt file with information about the samples 

#Prepare workspace 

Load libraries
```{r message= FALSE, warning=FALSE}
#install.packages("ggplot2")
#install.packages("dplyr")
#install.packages("tidyr")
#devtools::install_github("sjmgarnier/viridis")
#install.packages("VennDiagram")
#install.packages("rtracklayer")

#source("https://bioconductor.org/biocLite.R")
#biocLite("limma")
#biocLite("edgeR")
#biocLite("DEXSeq")
#biocLite("GenomicRanges")
#biocLite("Rsubread")



library(limma)
library(edgeR)
library(ggplot2)
library(dplyr)
library(plyr)
library(tidyr)
library(DEXSeq)
library(GenomicRanges)
library(rtracklayer)
library(Rsubread)
library(viridis) #colour blind friendly palette for plots
library(VennDiagram)
library(rtracklayer)
library(dplyr)
library(data.table)
library(gplots)
library(UpSetR)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Gene orthology
##The datafiles: psl and gtf

This section of the script uses 2 types of files: 
 
 1. **Psl**: output from blat 
         Blat done with genes assembled with StringTie in a non-model species
         matched to a model species (e.g. _Apis mellifera_)

* **MSTRG** -> StringTie genes in the non-model
   
* **GB** -> Honey bee genes in the model species

 2. **Gtf**: gtf file from the model species 
         
Contains the gene ids from FlyBase FB also the gene names (e.g. dsx)

**Goal:** 

    MSTRG#(non-model) -> GB#(model) -> gene_name(model)


Read in the psl file (output file from blat)
```{r}
dataFile = read.table('/Users/afarre/Unimelb/Doublesex/Mpharaonis/necklace_ame2/relST_genomeST.psl', header = FALSE, sep = "\t", skip = 5, stringsAsFactors = F)
```

Psl file has no header. Create vector of column names
```{r}
names_psl <- unlist(strsplit("matches
misMatches
repMatches
nCount
qNumInsert
qBaseInsert
tNumInsert
tBaseInsert
strand
qName
qSize
qStart
qEnd
tName
tSize
tStart
tEnd
blockCount
blockSizes
qStarts
tStarts", "\n"))
```

Change names of the psl object. 

Columns with query (non-model species) and target (model species) gene names:
```{r cols.print=3, rows.print=3}
names(dataFile) <- names_psl
dplyr::select(dataFile[14:18,], qName, tName)
```

Read in the gtf file 
```{r rows.print=5}
dme_gtf <- import.gff3("/Users/afarre/Unimelb/Doublesex/Amellifera/GCF_000002195.4_Amel_4.5_genomic.gff")
as.data.frame(dme_gtf)

dme_gtf$dbGeneIDlist <- paste(dme_gtf$Dbxref, sep=" ", collapse = ",")
```

##From the genes of the non-model organism to the model's

Using the psl file we get the matches of the StringTie genes to the FlyBase ones

Get all stringtie gene names in a new column
```{r rows.print=5}
dataFile$qmstrg <- sub(".*(MSTRG\\.\\d*).*", "\\1", dataFile$qName)
as.data.frame(unique(dataFile$qmstrg))
```

Get all FlyBase (FB*) gene ids in a list, each entry corresponds to the FB gene match
```{r results="hold"}
dataFile$ID <- sub("^(gene\\d*):.*", "\\1", dataFile$tName)

dataFile %>% 
  dplyr::select(ID, tName, qName, qmstrg)
```

##From gene IDs to gene names in _Apis mellifera_

Using the gtf file 

For each gene, find the gene name using the FB gene id
```{r rows.print=5}
mcols(dme_gtf)$geneName <- sub(".*(gene\\d*).*", "\\1", mcols(dme_gtf)$product)
```

For each StringTie gene in the Query (non-model) species find the match in the Related (model) species
```{r results="hold"}

ame_metaData <- as.data.frame(mcols(dme_gtf))

ame_metaData$geneName <- sub("(.*), transcript variant .*", "\\1", ame_metaData$product)

ame_metaDataRNA <- ame_metaData %>%
  dplyr::filter(grepl("rna*", ID)) %>% #keep only rows that match rna* in the ID column
  dplyr::select(type, ID, Name, gene, geneName) %>% #keep only certain columns (others have NA and uninteresting stuff)
  dplyr::group_by(gene, geneName) %>% #group by gene and geneName 
  dplyr::filter(row_number(ID) == 1) %>% #keep only one entry per gene
  dplyr::select(gene, geneName) 

ameMatches <- ame_metaData %>%
  dplyr::filter(grepl("gene*", ID)) %>% #keep only rows that match gene* in the ID column
  dplyr::select(type, ID, gene, dbGeneIDlist) %>% #keep only certain columns (others have NA and uninteresting stuff)
  merge(ame_metaDataRNA, by="gene",  all = TRUE)%>%
  merge(dataFile, by="ID") %>%
  dplyr::select(ID, qmstrg, type, gene, geneName, dbGeneIDlist) %>%
  unique()

ameMatches$dbGeneID <- sub("BEEBASE:(GB\\d*).*", "\\1", ameMatches$dbGeneIDlist)
```


#Differential Splicing: Analysis of necklace output of _Monomorium pharaonis_ 
##Pre-processing
###Read in the data info

DexSeq requires a table with information about the samples as well as the raw counts for each exon in each sample
And grouping information for which exons belong to which genes

Read in the table with all the information of the SRA samples

```{r rows.print=5}
sample_id<-read.table("/Users/afarre/Unimelb/Doublesex/Mpharaonis/data_info/Mph_pe_SRR_Acc_List.txt", col.names = "sampleID", stringsAsFactors = F )
sample_id 

info_samples<-read.table("/Users/afarre/Unimelb/Doublesex/Mpharaonis/data_info/Mph_pe_SraRunTable.txt", sep = "\t", header = T,  stringsAsFactors = F )

info_samples <- info_samples[info_samples$BioProject_s=="PRJDB4088",]
info_samples
```

Create the data_info table for DEXseq

```{r rows.print=10}
#Get the sample names per caste 
Queens <- info_samples[grep(".*Q\\d$", info_samples$sample_name_s, perl = T),]$Run_s
Workers <- c(info_samples[grep(".*W\\d$", info_samples$sample_name_s, perl = T),]$Run_s, info_samples[grep("^PHARAO\\..*COL\\d$", info_samples$sample_name_s, perl = T),]$Run_s)

DRP002877 <- info_samples[grep(".*[Q|W]\\d$", info_samples$sample_name_s, perl = T),]$Run_s

#add a population and caste column
info_samples$casteName <- ""
info_samples$caste <- ""

info_samples[info_samples$Run_s %in% Queens,]$casteName <- "Queen"
info_samples[info_samples$Run_s %in% Workers,]$casteName <- "Worker"

info_samples[info_samples$Run_s %in% Queens,]$caste <- "1"
info_samples[info_samples$Run_s %in% Workers,]$caste <- "2"


data_info <- dplyr::select(info_samples, Run_s, casteName)
row.names(data_info) <- data_info$Run_s
data_info$casteName <- as.factor(data_info$casteName)
colnames(data_info) <- c("sample", "condition")

```


###Read in the count data
Read in the data (count matrix)
```{r  rows.print=5}
#First get counts table
counts <- read.table("/Users/afarre/Unimelb/Doublesex/Mpharaonis/necklace_ame2/block.counts",sep="\t",header=TRUE,stringsAsFactors=F)
names(counts) <- gsub(".*(DRR.*)\\.bam","\\1", names(counts))

#select only the samples that we want to use
#in this case it's only the PE samples
counts <- dplyr::select(counts, Geneid, Chr, Start, End, Strand, Length, data_info$sample)
counts$Geneid <- gsub(":","_",counts$Geneid)

```

Create the exon_ids column with a unique identifier for each exon

```{r rows.print=5}
#each exon needs a unique name
#create a column with unique exon ids 
#exonid = geneid + exon number
counts <- do.call(rbind, lapply(unique(counts$Geneid), function(x){ 
  mutate(dplyr::filter(counts,Geneid==x), exon_ids=
           paste(dplyr::filter(counts,Geneid==x)$Geneid, 
                 rownames(dplyr::filter(counts,Geneid==x)), sep = "."))
} ))

row.names(counts) <- counts$exon_ids
```

Convert the counts data frame into a matrix (i.e exclude first 6 columns: gene, chr, etc.)

```{r}
#create the matrix with only the counts
count_matrix <- data.matrix(dplyr::select(counts, -Geneid, -Chr, -Start, -End, -Strand, -Length, -exon_ids))
count_matrix <- round(count_matrix)
# gene_ids <- counts$Geneid
# exon_ids <- counts$exon_ids
```

###Create the dge object

Create the dge object with the conts matrix and the gene id information

Counts is a matrix of exon-level counts, and GeneID identities which gene each exon belongs

```{r results="hold"}
dge <- DGEList(counts=count_matrix, genes=counts$Geneid)
dge$genes$GeneID <- counts$Geneid
dge
```

###Create the design matrix

Create a design matrix

First asign numbers to each caste 

Then create the design matrix for limma 

```{r rows.print=10}
data_infoLimma <- dplyr::select(info_samples, Run_s, caste)
row.names(data_infoLimma) <- data_infoLimma$Run_s
data_infoLimma$caste <- as.factor(data_infoLimma$caste)

design <- model.matrix(~ 0 + data_infoLimma$caste)
row.names(design) <- data_infoLimma$Run_s

#order matrix by row names and change column names to match condition (i.e queen, worker, etc)
design <- design[order(rownames(design)),]
colnames(design)<- c("queen", "worker")

as.data.frame(design)

```

Include caste and population info into the DGE object.

```{r rows.print=10}
dge$samples$caste <- ""
dge$samples[rownames(dge$samples) %in% rownames(design[as.data.frame(design)$queen==1,]),]$caste <- "1"
dge$samples[rownames(dge$samples) %in% rownames(design[as.data.frame(design)$worker==1,]),]$caste <- "2"

dge$samples
```

###Calculate normalized library sizes, generate basic descriptive plots

```{r}
dge <- calcNormFactors(dge)
```

Some barplots about mapping stats.

"Raw" library sizes:
```{r}
lib.size <- data.frame(lib.size=sort(colSums(dge$counts)), sample=names(sort(colSums(dge$counts))), stringsAsFactors = F)

lib.size$caste <- data_infoLimma[lib.size$sample,]$caste

lib.size$caste <- gsub(pattern = "1", replacement = "Queen", x = lib.size$caste)
lib.size$caste <- gsub(pattern = "2", replacement = "Worker", x = lib.size$caste)

ggplot(lib.size, aes(sample, lib.size, fill=caste))+
  geom_bar(stat="identity")+
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10))

```


Normalised library sizes

```{r}
norm.lib.size <- dge$samples
norm.lib.size$sample <- rownames(norm.lib.size) 
norm.lib.size$norm.lib.size <- norm.lib.size$lib.size * norm.lib.size$norm.factors
norm.lib.size <- norm.lib.size[order(norm.lib.size$norm.lib.size),]

norm.lib.size$caste <- gsub(pattern = "1", replacement = "Queen", x = norm.lib.size$caste)
norm.lib.size$caste <- gsub(pattern = "2", replacement = "Worker", x = norm.lib.size$caste)

ggplot(norm.lib.size, aes(sample, norm.lib.size, fill=caste))+
  geom_bar(stat="identity")+
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10))

```

####Number of expressed genes per sample

Plot the number of genes expressed in each sample

```{r}
ex_gene_sample <- apply(count_matrix, 2, function(x){ sum(x!=0) })
ex_gene_sample <- as.data.frame(ex_gene_sample, stringsAsFactors = F)
names(ex_gene_sample) <- "Num_expressed_genes"
ex_gene_sample$sample <- rownames(ex_gene_sample)

ex_gene_sample$caste <- data_infoLimma[ex_gene_sample$sample,]$caste

ex_gene_sample$caste <- gsub(pattern = "1", replacement = "Queen", x = ex_gene_sample$caste)
ex_gene_sample$caste <- gsub(pattern = "2", replacement = "Worker", x = ex_gene_sample$caste)
ex_gene_sample$caste <- gsub(pattern = "3", replacement = "Male", x = ex_gene_sample$caste)

ggplot(ex_gene_sample, aes(sample, Num_expressed_genes, fill=caste ))+
  geom_bar(stat="identity")+
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 10))

```

####Number of expressed genes in % samples

For each gene calculate the % of samples in which it is expressed. 

Calculate how many genes are expressed in each % interval of samples.

```{r}

genes_percent_samples <- as.data.frame(apply(count_matrix, 1, function(x){ sum(x!=0)/dim(count_matrix)[2]*100 }))
names(genes_percent_samples) <- "Percent_samples"
genes_percent_samples$gene <- counts$Geneid


ggplot(genes_percent_samples, aes(Percent_samples) )+
  geom_histogram(stat="bin", binwidth=5, aes(fill=..count..))+
  scale_fill_viridis() +
  theme_bw()

```


###Calculate cpm, Normalize and filter

Different filtering either by 

* counts>10 

* cpm>1 in n=3

* cpm>1 in n=(n caste)/2

```{r}
#A <- rowSums(dge$counts)
#dge <- dge[A>10, , keep.lib.sizes=FALSE]

#isexpr <- rowSums(cpm(dge) > 1) >=3
#dge <- dge[isexpr,,keep.lib.sizes=FALSE]

#dge <- calcNormFactors(dge)
```


Calculate log2 CPM
```{r}
cpmNorm <- cpm(dge, normalized.lib.sizes=TRUE, log=TRUE, prior.count=0.25) 
plot.new()
plotDensities(cpmNorm, group = norm.lib.size[colnames(cpmNorm),]$caste) 
abline(v=0)
```

Filter:

if n caste <3 then cpm>1 in n=(n caste)

if n caste >3 then cpm>1 in n=(n caste)/2

```{r}
cpm.dge <- dge
cpm.dge$counts <- cpmNorm

#create a normalization factor 
#if #samples < 3 then norm.factor=1
#if #samples > 3 then norm.factor=0.5

norm.samples <- data.frame(caste=levels(as.factor(cpm.dge$samples$caste)), stringsAsFactors = F)
norm.samples$norm <- ""
isexpr <- unlist(lapply(levels(as.factor(cpm.dge$samples$caste)), function(x){
  if(sum(cpm.dge$samples$caste==x)<3){
    1
  }
  else{
    0.5
  }}))
norm.samples$norm <- isexpr

#the normalization factor is used in
# if n caste <3 then cpm>1 in n=(n caste)
# if n caste >3 then cpm>1 in n=(n caste)/2

#which means that we * by the normalization factor 

#cpm>1
dge.filt1 <- dge[
    rowSums(cpm.dge$counts[,rownames(cpm.dge$samples[cpm.dge$samples$caste=="1",])] >1) >= (sum(cpm.dge$samples$caste=="1")*norm.samples$norm[norm.samples$caste=="1"])|
    rowSums(cpm.dge$counts[,rownames(cpm.dge$samples[cpm.dge$samples$caste=="2",])] >1) >= (sum(cpm.dge$samples$caste=="2")*norm.samples$norm[norm.samples$caste=="2"])
    ,,keep.lib.sizes=FALSE]

cat(dim(dge.filt1)[1], "exons kept in",dim(dge.filt1)[2],"samples if the filter is set to cmp>1\n")

#cmp>0
dge.filt0 <- dge[
    rowSums(cpm.dge$counts[,rownames(cpm.dge$samples[cpm.dge$samples$caste=="1",])] >0) >= (sum(cpm.dge$samples$caste=="1")*norm.samples$norm[norm.samples$caste=="1"])|
    rowSums(cpm.dge$counts[,rownames(cpm.dge$samples[cpm.dge$samples$caste=="2",])] >0) >= (sum(cpm.dge$samples$caste=="2")*norm.samples$norm[norm.samples$caste=="2"])
    ,,keep.lib.sizes=FALSE]

cat(dim(dge.filt0)[1], "exons kept in",dim(dge.filt0)[2],"samples if the filter is set to cmp>0")

```

Recalculate lib size:

```{r rows.print=5}
# dge.filt1 <- calcNormFactors(dge.filt1)
dge.filt0 <- calcNormFactors(dge.filt0)
info_counts <- dplyr::select(counts, Geneid, Chr, Start, End, Strand, Length, exon_ids)
info_counts_filt0 <- info_counts[info_counts$exon_ids %in% row.names(dge.filt0$counts),]
dge.filt0$genes <- cbind(dge.filt0$genes, info_counts_filt0)
dge.filt0$genes <- dplyr::select(dge.filt0$genes, -Geneid, -Chr)
dge.filt0$genes 
```

##DEXSeq

```{r}
count_matrix <- data.matrix(dge.filt0$counts)
gene_ids <- dge.filt0$genes$GeneID
exon_ids <- dge.filt0$genes$exon_ids


#design
dxd <- DEXSeqDataSet(count_matrix,design=~sample + exon + condition:exon, featureID=as.factor(exon_ids),
                     groupID=as.factor(gene_ids),sampleData=data_info)

## Could use the DEXSeqDataSetFromFeatureCounts function, but never got it to work...
# flattenedFile <- import.gff("/Users/afarre/Unimelb/Doublesex/Mpharaonis/necklace_ame/SuperDuper.gff")
# source("/Users/afarre/Unimelb/Doublesex/R/load_SubreadOutput.R")
# dxd2 <- DEXSeqDataSetFromFeatureCounts(countfile = "/Users/afarre/Unimelb/Doublesex/Mpharaonis/necklace_ame/block_counts.out" ,design=~sample + exon + condition:exon, sampleData=data_info, flattenedfile="/Users/afarre/Unimelb/Doublesex/Mpharaonis/necklace_ame/SuperDuper.gff")

##Estimate size factors and dispersions (DEXseq does this based on a negative bionmial distribution

dxd <- estimateSizeFactors(dxd)
dxd <- estimateDispersions(dxd)

#Test for DEU
dxd <- testForDEU(dxd)

#Extract the results
res <- DEXSeqResults(dxd)

#Get p-value per gene based on the p-values for all exons in the gene/cluster
pgq <- perGeneQValue(res, p = "pvalue" ) 
pgq

## Save results to a text file and R object
tmp <- cbind(gene = names(pgq), "adjP" = pgq)
```

```{r}
adjPperGene <- as.data.frame(tmp, stringsAsFactors=FALSE)
adjPperGene$adjP <- as.numeric(adjPperGene$adjP)
signif.diff.caste <- adjPperGene[adjPperGene$adjP<0.05,]

adjPperGene[adjPperGene$gene=="gene8203_MSTRG.9358",]

```

Save the Rdata and tables 

```{r}
save(dxd, res, pgq, file = "/Users/afarre/Unimelb/Doublesex/R/DEXSeq/Mph_DEXseq.Rdata")
write.table(tmp, file = "/Users/afarre/Unimelb/Doublesex/R/DEXSeq/Mph_DEXseq_byGene.txt", col.names = TRUE, row.names = FALSE, quote = FALSE, sep = "\t")
write.table(DEXSeqResults(dxd), file="/Users/afarre/Unimelb/Doublesex/R/DEXSeq/Mph_DEXseq_byExon.txt", col.names = TRUE, row.names = FALSE, quote = FALSE, sep = "\t" )
```

##Differential splicing
###Differentially spliced exons by caste


```{r rows.print=10}

#then use the function topSplice with the coeficient (comparision) that we're interested in
signif.diff.caste$qmstrg <- sub(".*(MSTRG\\.\\d*).*", "\\1", signif.diff.caste$gene)
signif.diff.caste

#Find the names of the genes is Apis 
diffMatches.caste <- signif.diff.caste %>%
  merge(ameMatches, by="qmstrg") 


int_genes_ds <- diffMatches.caste[grep(".*transformer.*", diffMatches.caste$geneName, ignore.case = TRUE),]
int_genes_ds <- rbind(int_genes_ds, diffMatches.caste[grep(".*vitellogenin.*", diffMatches.caste$geneName, ignore.case = TRUE),])
int_genes_ds <- rbind(int_genes_ds, diffMatches.caste[grep(".*bric.*", diffMatches.caste$geneName, ignore.case = TRUE),])
int_genes_ds <- rbind(int_genes_ds, diffMatches.caste[grep(".*juvenile hormone.*", diffMatches.caste$geneName, ignore.case = TRUE),])
int_genes_ds <- rbind(int_genes_ds, diffMatches.caste[grep(".*insulin.*", diffMatches.caste$geneName, ignore.case = TRUE),])
int_genes_ds <- rbind(int_genes_ds, diffMatches.caste[grep(".*royal.*", diffMatches.caste$geneName, ignore.case = TRUE),])
int_genes_ds <- rbind(int_genes_ds, diffMatches.caste[grep(".*forkhead.*O-like.*", diffMatches.caste$geneName, ignore.case = TRUE),])
int_genes_ds <- rbind(int_genes_ds, diffMatches.caste[grep(".*yellow.*", diffMatches.caste$geneName, ignore.case = TRUE),])
int_genes_ds <- rbind(int_genes_ds, diffMatches.caste[grep(".*doublesex.*", diffMatches.caste$geneName, ignore.case = TRUE),])
int_genes_ds <- rbind(int_genes_ds, diffMatches.caste[grep(".*fruitless.*", diffMatches.caste$geneName, ignore.case = TRUE),])
dplyr::select(int_genes_ds, gene.y,geneName, adjP)


write.table(diffMatches.caste,"/Users/afarre/Unimelb/Doublesex/R/DEXSeq/dsCaste_dexseq_mph.txt", col.names = TRUE)


```

Fruitless: 
```{r}
# fru <- diffMatches.caste[grep(".*fruitless.*", diffMatches.caste$geneName),]$GeneID
# 
# plotSplice(ex.caste.loess, coef = 1, geneid = fru)
# 
# gene_ids[grep("MSTRG.9358",gene_ids )]
# 
# 
# res[grep("MSTRG.9358",res$groupID ),]
# plotDEXSeq(res, "MSTRG.10237")

```

###Differential splicing of _doublesex_ by caste

Which will show individual exons that are enriched or depleted relative to other exons in the same gene. 

To display the pattern of exons in the top genes.

Plot doublesex (identified by blast beforehand)  "gene8203:MSTRG.9358"

```{r}
# par(mar=c(1,1,1,1))
# layout( matrix(c(1,2),ncol=1), heights = c(1,3))
# #layout.show(n = 6)
# plot.new()
# text(0.5,0.5,"Queen-Worker",cex=2,font=2 )
# plotSplice(ex.caste.loess, coef = 1, geneid = "gene8203:MSTRG.9358")
# 

#dev.off()

```

P values of the exons per contrast

```{r}
# 
# #diff$genes=="gene13749:MSTRG.15287"
# pvalues_dsx <- ex.caste.loess$p.value[ex.caste.loess$gene.firstexon["gene8203:MSTRG.9358"]:ex.caste.loess$gene.lastexon["gene8203:MSTRG.9358"],]
# # rownames(pvalues_dsx) <- c("exon1","exon3","exon5","exon6","exon7")
# pvalues_dsx
# # ex.caste.loess[ex.caste.loess$gene.firstexon["gene8203:MSTRG.9358"]:ex.caste.loess$gene.lastexon["gene8203:MSTRG.9358"],]
# # dge[dge$genes$gene=="gene8203:MSTRG.9358",]
```


